rfc/rfc-83/rfc-83.md

```diff
@@ -0,0 +1,233 @@
+When there are many historical partitions, such as 640,000 in our production environment, this scanning and planning operation becomes very inefficient. 
+For Flink, it often leads to checkpoint timeouts, resulting in data delays.
+As for cleaning, we already have the ability to do cleaning for incremental partitions.
+
+This RFC will draw on the design of Incremental Clean to generalize the capability of processing incremental partitions to all table services, such as Clustering and Compaction.
+
+## Background
+
+`earliestInstantToRetain` in clean plan meta
+
+HoodieCleanerPlan.avsc
+
+```text
+{
+  "namespace": "org.apache.hudi.avro.model",
+  "type": "record",
+  "name": "HoodieCleanerPlan",
+  "fields": [
+    {
+      "name": "earliestInstantToRetain",
+      "type":["null", {
+        "type": "record",
+        "name": "HoodieActionInstant",
+        "fields": [
+          {
+            "name": "timestamp",
+            "type": "string"
+          },
+          {
+            "name": "action",
+            "type": "string"
+          },
+          {
+            "name": "state",
+            "type": "string"
+          }
+        ]
+      }],
+      "default" : null
+    },
+    xxxx
+  ]
+}
+```
+
+`EarliestCommitToRetan` in clean commit meta
+
+HoodieCleanMetadata.avsc
+
+```text
+{"namespace": "org.apache.hudi.avro.model",
+ "type": "record",
+ "name": "HoodieCleanMetadata",
+ "fields": [
+     xxxx,
+     {"name": "earliestCommitToRetain", "type": "string"},
+     xxxx
+ ]
+}
+```
+How to get incremental partitions during cleaning
+
+![cleanIncrementalpartitions.png](cleanIncrementalpartitions.png)
+
+**Note**
+`EarliestCommitToRetain` is recorded in `HoodieCleanMetadata`
+newInstantToRetain is computed based on Clean configs such as `hoodie.clean.commits.retained` and will be record in clean meta as new EarliestCommitToRetain
+
+## Design And Implementation
+
+### Changes in TableService Metadata Schema
+
+Add new column `earliestInstantToRetain` (default null) in Clustering/Compaction plan same as `earliestInstantToRetain` in clean plan
+
+```text
+    {
+      "name": "earliestInstantToRetain",
+      "type":["null", {
+        "type": "record",
+        "name": "HoodieActionInstant",
+        "fields": [
+          {
+            "name": "timestamp",
+            "type": "string"
+          },
+          {
+            "name": "action",
+            "type": "string"
+          },
+          {
+            "name": "state",
+            "type": "string"
+          }
+        ]
+      }],
+      "default" : null
+    },
+```
+
+We also need a unified interface/abstract-class to control the Plan behavior of the TableService including clustering and compaction.
+
+### Abstraction
+
+Use `PartitionBaseTableServicePlanStrategy` to control the behavior of getting partitions, filter partitions and generate table service plan etc.
```

Comment:
Maybe we name it `IncrementalPartitionAwareStrategy` to emphasize it is "incremental".

Reply:
changed